Data/IP/Addr.hs

```diff
@@ -258,85 +259,186 @@ showIPv6 ip@(IP6 (a1,a2,a3,a4))
 --
 
 {-|
-  The 'toIPv4' function takes a list of 'Int' and returns 'IPv4'.
+  The 'toIPv4' function returns the 'IPv4' address corresponding to the given
+  list of 'Int' octets.  The function is strict in the four elements of the
+  list.  An error is returned if the list has a differnet length.  The input
+  elements are silently truncated to their 8 least-significant bits before they
+  are combined to form the IPv4 address.
 
 >>> toIPv4 [192,0,2,1]
 192.0.2.1
 -}
 toIPv4 :: [Int] -> IPv4
-toIPv4 = IP4 . toWord32
+toIPv4 [a1, a2, a3, a4] = IP4 w
   where
-    toWord32 [a1,a2,a3,a4] = fromIntegral $ shift a1 24 + shift a2 16 + shift a3 8 + a4
-    toWord32 _             = error "toWord32"
+    !w = (fromIntegral a1 .&. 0xff) `unsafeShiftL` 24 .|.
+         (fromIntegral a2 .&. 0xff) `unsafeShiftL` 16 .|.
+         (fromIntegral a3 .&. 0xff) `unsafeShiftL`  8 .|.
+         (fromIntegral a4 .&. 0xff)
+toIPv4 _ = error "IPv4 field list length != 4"
+{-# INLINE toIPv4 #-}
 
 {-|
-  The 'toIPv6' function takes a list of 'Int' and returns 'IPv6'.
+  The 'toIPv4w' function constructs the 'IPv4' address corresponding to the
+  given 'Word32' value.  Unlike the 'fromHostAddress' function, it is strict in
+  the input value, which here is in host byte order.
+
+>>> toIPv4w 0xc0000201
+192.0.2.1
+
+@since 1.7.9
+-}
+toIPv4w :: Word32 -> IPv4
+toIPv4w !w = IP4 w
+{-# INLINE toIPv4w #-}
+
+{-|
+  The 'toIPv6' function returns the 'IPv6' address corresponding to the given
+  list of eight 16-bit 'Int's.  The function is strict in the eight elements of
+  the list.  An error is returned if the list has a differnet length.  The
+  input elements are in host byte order and are silently truncated to their 16
+  least-signicant bits before they are combined to form the IPv6 address.
 
 >>> toIPv6 [0x2001,0xDB8,0,0,0,0,0,1]
 2001:db8::1
 -}
 toIPv6 :: [Int] -> IPv6
-toIPv6 ad = IP6 (x1,x2,x3,x4)
+toIPv6 [i1,i2,i3,i4,i5,i6,i7,i8] = IP6 (x1,x2,x3,x4)
   where
-    [x1,x2,x3,x4] = map toWord32 $ split2 ad
-    split2 [] = []
-    split2 x  = take 2 x : split2 (drop 2 x)
-    toWord32 [a1,a2] = fromIntegral $ shift a1 16 + a2
-    toWord32 _       = error "toWord32"
+    !x1 = fromIntegral $ (i1 .&. 0xffff) `unsafeShiftL` 16 .|. (i2 .&. 0xffff)
```

Comment:
Again, `!`s are not necessary.

Reply:
No, actually, these are necessary, because tuples are not strict in their elements and the output is a 4-tuple.